src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,183 @@
+// Copyright (c) 2010 Jack Palevich. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+//    * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 declare(strict_types=1);
 
 /**
- * UPnP port forwarding support. Only for Windows
+ * UPnP port forwarding support.
  */
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string|null */
+	private static $serviceURL;
+	
+	public static function getControlUrl() : string{
+		try{
+			$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+			socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, ["sec" => 3, "usec" => 0]);
+			$contents =
+				"M-SEARCH * HTTP/1.1\r\n" .
+				"MX: 2\r\n" .
+				"HOST: 239.255.255.250:1900\r\n" .
+				"MAN: \"ssdp:discover\"\r\n" .
+				"ST: upnp:rootdevice\r\n\r\n";
+			socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+			socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+			socket_close($socket);
+		}catch(\Exception $e){
```

Comment:
Not a fan of global `Exception` caught here.

Reply:
I need to put `set_error_handler([Utils::class, 'errorExceptionHandler']);` before try and catch `ErrorException` and put `restore_error_handler();` in finally, right?

Reply:
No, it's already set by default. That said, it's better to use @ and check for false returns, because static analysers can't make sense of error handlers.